Public key pinning

rclee opened the pull request:
Updating public key pinning to use new hashes and configuration endpoint.

rclee committed:
Fix race condition in cedar test

"created" NSDate was causing a failure some of the time when the date time was off by a second.

rclee committed:
Remove hard-coded date string

akitchen committed:
Allow users to opt into a custom configuration profile

Co-authored-by: Randall Lee <randall.lee@mapbox.com>

akitchen committed:
Checkpoint adding additional functionality to custom profile

Co-authored-by: Randall Lee <randall.lee@mapbox.com>

rclee committed:
Add CN and Com hashes to project

rclee committed:
Add CN and Com hashes to project

akitchen committed:
Refactor MMEEventsConfiguration to handle its own instantiation

Further refactoring of object creation in tests might help us here
 - Fewer singletons would mean less statefulness to worry about
 - Avoiding mutability will help us avoid accidental bugs in future

Co-authored-by: Randall Lee <randall.lee@mapbox.com>

akitchen committed:
Added .idea to .gitignore

rclee committed:
Add blacklist HTTP request method

-need to persist list
-focus on renaming of methods

rclee committed:
Add blacklist HTTP request method

-need to persist list
-focus on renaming of methods

rclee committed:
Merge branch 'custom-profile' into develop

Bringing in refactoring changes for custom configuration profiles

Co-Authored-By: Andrew Kitchen <akitchen@gmail.com>

rclee committed:
Rebase on develop branch

rclee committed:
Rebase on master

Co-authored-by: Randall Lee <randall.lee@mapbox.com>

Checkpoint adding additional functionality to custom profile

Co-authored-by: Randall Lee <randall.lee@mapbox.com>

Add CN and Com hashes to project


Remove hard-coded date string
Add blacklist HTTP request method

-need to persist list
-focus on renaming of methods

Fix race condition in cedar test

"created" NSDate was causing a failure some of the time when the date time was off by a second.
Add CN and Com hashes to project


Add blacklist HTTP request method

-need to persist list
-focus on renaming of methods

rclee committed:
Refactored TrustKit initialization to avoid the singleton

This will allow us to reconfigure as needed at run time

Co-authored-by: Andrew Kitchen <andrew.kitchen@mapbox.com>

rclee committed:
Refactored TrustKit initialization to avoid the singleton

This will allow us to reconfigure as needed at run time

Co-authored-by: Andrew Kitchen <andrew.kitchen@mapbox.com>

rclee committed:
Rename TrustKit helper class

Co-Authored-By: Andrew Kitchen <akitchen@gmail.com>

rclee committed:
Rename TrustKit helper class

Co-Authored-By: Andrew Kitchen <akitchen@gmail.com>

rclee committed:
Refactor MMEAPIClient

This will move http status error code into a helper method

Co-Authored-By: Andrew Kitchen <akitchen@gmail.com>

rclee committed:
Merge remote-tracking branch

rclee committed:
Create blacklist parsing utility

rclee committed:
Add MMEHashProvider

Should provide hashes to TrustKit. Also includes helper method to filter out blacklisted hashes.

rclee committed:
Add MMEHashProvider

Should provide hashes to TrustKit. Also includes helper method to filter out blacklisted hashes.

rclee committed:
Update TrustKit to use HashProvider

Injects MMEEventsConfiguration in order to gain access to blacklist.

rclee committed:
Update TrustKit to use HashProvider

Injects MMEEventsConfiguration in order to gain access to blacklist.

rclee committed:
Add MMEConfigurationUpdater and protocol

For now the configuration updater will only handle retrieval of revoked keys. In the future, retrieved configurations will contain more options.

rclee committed:
Setup MMEConfigurationUpdaterDelegate in MMEEventsManager

Calls for a configuration update as often as a turnstile is sent. Reconfigure methods added to needed classes and are triggered by the delegate call.

rclee committed:
Rename config API method

rclee committed:
Only reconfigure during set intervals

Will attempt to reconfigure based on our API every ~24 hours unless configured otherwise.

rclee committed:
Reconfigure properties when required

rclee committed:
Merge branch 'master' into public-key-pinning

julianrex commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -66,6 +74,34 @@ - (void)setBaseURL:(NSURL *)baseURL {
+        NSString *reason = [NSString stringWithFormat:reasonFormat, (long)httpResponse.statusCode];
+        NSDictionary *userInfo = @{NSLocalizedDescriptionKey: description,
+                                   NSLocalizedFailureReasonErrorKey: reason};
+        statusError = [NSError errorWithDomain:MMEErrorDomain code:1 userInfo:userInfo];
```
I'd prefer to see an enum or similar for the error code here. It seems likely that we'd have other codes under the same `MMEErrorDomain`.

It might also be worth making it easier to access the HTTP status code from the info dictionary. As it stands you'd have to extract it from a format string. Consider adding the `NSHTTPURLResponse` to the dictionary.

julianrex commented on MapboxMobileEvents/MMEConfigurationUpdater.h:
```diff
@@ -0,0 +1,20 @@
+
+@protocol MMEConfigurationUpdaterDelegate <NSObject>
+
+- (void)configurationDidUpdate:(MMEEventsConfiguration *)configuration;
```
Delegate methods (by convention) have a parameter for the sender. So (and I struggle with naming delegates), perhaps something like:
```
- (void)configurator:(MMEConfigurationUpdater *)updater didUpdate:(MMEEventsConfiguration *)configuration;
```

julianrex commented on MapboxMobileEvents/MMEConfigurationUpdater.h:
```diff
@@ -0,0 +1,20 @@
+@interface MMEConfigurationUpdater : NSObject
+
+- (instancetype)init __attribute__((unavailable("This method is not available")));
+- (instancetype)initWithTimeInterval:(NSTimeInterval)timeInterval;
```
Perhaps use `NS_DESIGNATED_INITIALIZER` here?

julianrex commented on MapboxMobileEvents/MMEConfigurationUpdater.m:
```diff
@@ -0,0 +1,59 @@
+@interface MMEConfigurationUpdater ()
+
+@property (nonatomic) NSDate *configurationRotationDate;
+@property (nonatomic) MMEEventsConfiguration *configuration;
```
This should be a `copy` property.

julianrex commented on MapboxMobileEvents/MMEConfigurationUpdater.m:
```diff
@@ -0,0 +1,59 @@
+        [apiClient getConfigurationWithCompletionHandler:^(NSError * _Nullable error, NSData * _Nullable data) {
+            if (!error) {
+                self.configuration = [MMEEventsConfiguration configuration];
+                [self parseJSONFromData:data];
```
Could we combine these two calls into one?  e.g.
```
+[MMEEventsConfiguration configurationFromData:]
```

It's confusing as it is, and `parseJSONFromData` only modifies the configuration.

julianrex reviewed (commented):
Could this PR be broken into 2? 1 for the TrustKit changes and 1 for the configuration changes?

rclee committed:
Revert Trustkit cherry-picked commits

This reverts commit fd6bf9411b62090d01e88c1db39c99b3ff5ccce8.

Revert "Rename TrustKit helper class"

This reverts commit db82b0f8d55dc2b7411b0549c9a5e2847573a929.

Revert "Refactored TrustKit initialization to avoid the singleton"

This reverts commit c00f819089fa2d8c69c6c0202c282289b8f657b1.

Revert "Add china certificates and endpoint (#48)"

This reverts commit c7e6f657b3addef27db87496cf5870c5798d1dea.

rclee committed:
Revert "Add CN and Com hashes to project"

This reverts commit 0ad5a8189a6c434c0db9b43116e1bf38e7848b9a.

rclee committed:
Change hashes to readonly

rclee commented on MapboxMobileEvents/MMEEventsManager.m:
```diff
@@ -202,6 +212,8 @@ - (void)flush {
 }
 
 - (void)sendTurnstileEvent {
+    [self.configurationUpdater updateConfigurationFromAPIClient:self.apiClient];
```
Negative. There aren't any plans currently for configurations to affect events directly.

rclee commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -66,6 +74,34 @@ - (void)setBaseURL:(NSURL *)baseURL {
+    [request setValue:MMEAPIClientHeaderFieldContentTypeValue forHTTPHeaderField:MMEAPIClientHeaderFieldContentTypeKey];
+    [request setHTTPMethod:MMEAPIClientHTTPMethodGet];
+    
+    return [request copy];
```
I'm not sure honestly. I didn't add this line but it doesn't really seem necessary does it?

julianrex replied:
Yeah, I don't think you need it here.

rclee committed:
Move configuration parsing from updater

rclee committed:
Add HTTPURLResponse check

rclee committed:
Designate MMEConfigurationUpdater initializer

rclee committed:
Rename MMEConfigurationUpdater and delegate

rclee committed:
No longer make a copy of request

rclee committed:
Add error code enum and HTTP response to userInfo

rclee committed:
Merge branch 'trustkit-configuration' into public-key-pinning

Some methods in the new Trustkit configuration need to be merged for proper testing

julianrex commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -29,20 +33,15 @@ - (instancetype)initWithAccessToken:(NSString *)accessToken userAgentBase:(NSStr
     NSURLRequest *request = [self requestForEvents:events];
     [self.sessionWrapper processRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
-        NSError *statusError = nil;
         NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
```
You have another cast here - so same kind of check probably needs to go here?

rclee committed:
Fix merge issues

rclee committed:
Add error for an unexpected response type

Also, adds additional response check

akitchen commented on MapboxMobileEvents/MMETimerManager.m:
```diff
@@ -19,6 +19,8 @@ - (instancetype)initWithTimeInterval:(NSTimeInterval)timeInterval target:(id)aTa
 - (void)start {
     [self cancel];
+    // NOTE: this has the side-effect of scheduling the timer on the current or main run loop.
+    // Once delayed startup is implemented, the scheduling of this timer may also need to be delayed
```
Is there a plan to test this pattern in the field, based on this comment? We need to be careful about the memory management implications of using `NSTimer`. I like that it's encapsulated in a manager class which should make it easier to break any potential memory retain cycles, but nonetheless it's something we're going to want to put some additional thought into.

rclee committed:
Fix other merge issues

`copy` on MMEConfigurator -> MMEEventsConfiguration was causing a crash.

Add back China URLs

rclee committed:
Merge branch 'master' into public-key-pinning

rclee committed:
Add MMEHashProviderTests

rclee committed:
Fix Cedar tests

rclee committed:
Fix cedar test with missing response

rclee committed:
Merge branch 'master' into public-key-pinning

julianrex commented on MapboxMobileEvents/MMEEventsManager.m:
```diff
@@ -70,22 +76,28 @@ - (void)dealloc {
     
-    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(pauseOrResumeMetricsCollectionIfRequired) name:UIApplicationDidEnterBackgroundNotification object:nil];
-    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(pauseOrResumeMetricsCollectionIfRequired) name:UIApplicationDidBecomeActiveNotification object:nil];
+    void(^initialization)(void) = ^{
```
I think we should do the `weakself`/`strongself` dance here, and then return early if `strongself` is `nil`.  My concern is the timer manager at the bottom of the block and any potential retain cycles.

julianrex commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -181,6 +210,46 @@ - (NSURLRequest *)requestForEvents:(NSArray *)events {
+    
+    [httpBody appendData:[[NSString stringWithFormat:@"--%@--\r\n", boundary] dataUsingEncoding:NSUTF8StringEncoding]];
+    
+    NSLog(@"httpBody: %@",[[NSString alloc] initWithData:httpBody encoding:NSUTF8StringEncoding]);
```
We should remove this (or potentially replace with the telemetry log?)

rclee committed:
Add fixes

rclee committed:
Merge branch 'master' into public-key-pinning

julianrex commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -181,6 +213,49 @@ - (NSURLRequest *)requestForEvents:(NSArray *)events {
+    }
+    
+    NSString *mimetype = CFBridgingRelease(UTTypeCopyPreferredTagWithClass(UTI, kUTTagClassMIMEType));
+    if (mimetype == NULL) {
```
This `if` block isn't needed, since you return `mimetype` anyway. If it did return here, you'd also have an unreleased `UTI`.

rclee replied:
Good point

julianrex commented on MapboxMobileEvents/MMEEventsConfiguration.m:
```diff
@@ -13,22 +19,62 @@ + (instancetype)defaultEventsConfiguration {
+        CLLocationDistance customRadius = [customRadiusNumber isKindOfClass:[NSNumber class]] ? [customRadiusNumber doubleValue] : 1200.0;
+
+        id initializationDelayNumber = infoDictionary[kMMEStartupDelay];
+        NSTimeInterval initializationDelay = [initializationDelayNumber isKindOfClass:[NSNumber class]] ? [initializationDelayNumber doubleValue] : 10;
```
Same here for the `10`

rclee commented on MapboxMobileEvents/MMEAPIClient.m:
```diff
@@ -181,6 +213,49 @@ - (NSURLRequest *)requestForEvents:(NSArray *)events {
+        
+        [httpBody appendData:[[NSString stringWithFormat:@"--%@\r\n", boundary] dataUsingEncoding:NSUTF8StringEncoding]];
+        [httpBody appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"file\"; filename=\"%@\"\r\n", filename] dataUsingEncoding:NSUTF8StringEncoding]];
+        [httpBody appendData:[[NSString stringWithFormat:@"Content-Type: %@\r\n\r\n", mimetype] dataUsingEncoding:NSUTF8StringEncoding]];
```
It's actually accepted server-side without a mimetype at the moment.

rclee committed:
Add alterations

rclee committed:
Remove conditional on mimetype
